File: algo/house-robber-iii/rob.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def doRob(self, root):
        if not root:
            return [0, 0]
        left, right = self.doRob(root.left), self.doRob(root.right)
        return [max(left[0], left[1]) + max(right[0], right[1]), root.val + left[0] + right[0]]
    
    
    def rob(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        return max(self.doRob(root))
        

File: algo/house-robber-ii/rob.go
func max(a, b int) int {
    if a > b {
        return a
    } else {
        return b
    }
}

func doRob(nums []int) int {
    last, now := 0, 0
    for _, n := range nums {
        last, now = now, max(now, last+n)
    }
    return now
}


func rob(nums []int) int {
    if len(nums) == 0 {
        return 0
    }
    if len(nums) == 1 {
        return nums[0]
    }
    if len(nums) == 2 {
        return max(nums[0], nums[1])
    }
    return max(doRob(nums[:len(nums)-1]), doRob(nums[1:]))
}
